pkg/publisherstore/rstuf/rstuf.go

```diff
@@ -0,0 +1,124 @@
+// Copyright 2024 The Archivista Contributors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package rstuf
+
+import (
+	"bytes"
+	"context"
+	"crypto/sha256"
+	"encoding/hex"
+	"encoding/json"
+	"fmt"
+	"io"
+	"net/http"
+	"net/http/httputil"
+
+	"github.com/in-toto/archivista/pkg/config"
+	"github.com/sirupsen/logrus"
+)
+
+type RSTUF struct {
+	Host string
+}
+
+type Publisher interface {
+	Publish(ctx context.Context, gitoid string, payload []byte) error
+}
+
+func (r *RSTUF) parseRSTUFPayload(gitoid string, payload []byte) ([]byte, error) {
+	objHash := sha256.Sum256(payload)
+	// custom := make(map[string]any)
+	// custom["gitoid"] = gitoid
+	artifacts := []Artifact{
+		{
+			Path: gitoid,
+			Info: ArtifactInfo{
+				Length: len(payload),
+				Hashes: Hashes{
+					Sha256: hex.EncodeToString(objHash[:]),
+				},
+				// Custom: custom,
+			},
+		},
+	}
+
+	artifactPayload := ArtifactPayload{
+		Artifacts:         artifacts,
+		AddTaskIDToCustom: false,
+		PublishTargets:    true,
+	}
+
+	payloadBytes, err := json.Marshal(artifactPayload)
+	if err != nil {
+		return nil, fmt.Errorf("error marshaling payload: %v", err)
+	}
+	return payloadBytes, nil
+}
+
+func (r *RSTUF) Publish(ctx context.Context, gitoid string, payload []byte) error {
+	// this publisher allows integration with the RSTUF project to store
+	// the attestation and policy in the TUF metadata.
+	// this TUF metadata can be used to build truste when distributing the
+	// attestations and policies.
+	// Convert payload to JSON
+	url := r.Host + "/api/v1/artifacts"
+
+	payloadBytes, err := r.parseRSTUFPayload(gitoid, payload)
+	if err != nil {
+		return fmt.Errorf("error parsing payload: %v", err)
+	}
+
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(payloadBytes))
```

Comment:
I'm sure @kairoaraujo you'll know given the usage of RSTUF elsewher, but are there any issues hit when really large payloads are sent over?

And also come to think of it, would it be worth putting a context with a deadline in here? just a thought.

Reply:
The nice think is that we don't send the payload itself.
RSTUF just requires the size and the sha from the payload (dsse envelop)
https://github.com/in-toto/archivista/blob/7e31410c7050824adbdab7984564956a8ef9bb50/pkg/publisherstore/rstuf/rstuf.go#L44-L55